### DISCUSSION TITLE
当一个消息进入nonebot,  可以让nonebot等待一段时间(可以设置一个超时时间)， 等待第三方通过接口进入此次对话， 继续此次对话

### DISCUSSION THREAD

**Discussion by verydemo** [nonebot/discussions#483] [想法]
![image](https://github.com/user-attachments/assets/cedad2d4-09fe-4bb4-b77b-78d3298542b9)

**Comment by yanyongyu** [comment]
你可以使用event preprocessor hook来进行

**Reply by verydemo** [reply]
我现在通过redis实现的， 伪代码如下,   handle_image 在等待，  当外部触发了external_hook， handle_image 就往下走了
```
app = nonebot.get_app()

@app.post("/dockerimage/message")
async def external_hook(request: Request):
     await aioredis.rpush(message_id, content)
     

image = on_command("dockerimage", to_me())

@image.handle()
async def handle_image()
    value = await aioredis.blpop(msg_id, 60*3)
    await image.send(message=‘success’)
```



我试试你的方案,  把redis去掉看看

**Reply by verydemo** [reply]
好像用preprocessor 是在调用某个方法之前进行某一段逻辑， redis的blpop 还是得存在，或者用asyncio 实现一个异步阻塞队列

**Reply by yanyongyu** [reply]
不是很懂你的需求为什么要这么做，你可以在 event preprocessor 中等待你说的外部触发再返回。这样也能达到收到事件阻塞等待的效果。

**Reply by verydemo** [reply]
外部触发一般是http接口,  要怎么从这个接口跳转到 preprocessor 里面呢

**Comment by verydemo** [comment]
@yanyongyu 有相关的例子么， 能提供地址么？

**Reply by yanyongyu** [reply]
你可以异步等待接收到请求，你可以参考python官方文档里的 `asyncio.Future` 使用方法

**Reply by verydemo** [reply]
@yanyongyu 谢谢， 成功了
![image](https://github.com/user-attachments/assets/0a46a927-6cfe-4300-9b02-63fbf45a8adf)

**Reply by verydemo** [reply]
@yanyongyu 你好， 总感觉这样操作不优雅， 这种功能应该官方，或者第三方有实现才对，有更优雅的方法么？

**Reply by yanyongyu** [reply]
你想怎么优雅，你这个接口和事件处理本身就是完全独立的两个部分，通过异步future连接已经算是比较优雅的解决方法了
